File: backend/api/schedules.py
"""
Schedule API Endpoints for BIM Workbench

Provides endpoints for schedule management, quantity takeoff, and export functionality.
"""

from flask import Blueprint, request, jsonify, Response
from typing import List, Dict, Any
import json
from datetime import datetime

# Create blueprint
schedules_bp = Blueprint("schedules", __name__, url_prefix="/api/bim/schedules")


# ============================================================================
# Helper Functions
# ============================================================================


def calculate_area(obj: Dict) -> float:
    """Calculate surface area from object geometry (mm² to m²)"""
    width = obj.get("geometry", {}).get("width", 0)
    height = obj.get("geometry", {}).get("height", 0)
    if width and height:
        return (width * height) / 1_000_000
    return 0.0


def calculate_volume(obj: Dict) -> float:
    """Calculate volume from object geometry (mm³ to m³)"""
    width = obj.get("geometry", {}).get("width", 0)
    height = obj.get("geometry", {}).get("height", 0)
    thickness = obj.get("geometry", {}).get("thickness", 0)
    if width and height and thickness:
        return (width * height * thickness) / 1_000_000_000
    return 0.0


# ============================================================================
# Schedule Generation Endpoints
# ============================================================================


@schedules_bp.route("/generate", methods=["POST"])
def generate_schedule():
    """
    Generate a schedule from BIM objects.

    Request body:
    {
        "schedule_type": "door" | "window" | "material" | "custom",
        "objects": [list of BIM objects],
        "columns": [optional list of column definitions],
        "filters": [optional list of filters]
    }
    """
    data = request.get_json()

    if not data:
        return jsonify({"error": "No data provided"}), 400

    schedule_type = data.get("schedule_type")
    objects = data.get("objects", [])
    columns = data.get("columns", [])
    filters = data.get("filters", [])

    if not schedule_type:
        return jsonify({"error": "schedule_type is required"}), 400

    # Generate rows based on schedule type
    rows = []

    for obj in objects:
        obj_type = obj.get("type", "")

        # Apply filters
        include = True
        for f in filters:
            key = f.get("key")
            operator = f.get("operator", "equals")
            value = f.get("value")
            obj_value = obj.get(key, obj.get("properties", {}).get(key))

            if operator == "equals":
                if obj_value != value:
                    include = False
                    break
            elif operator == "contains":
                if value not in str(obj_value):
                    include = False
                    break
            elif operator == "gt":
                if float(obj_value) <= float(value):
                    include = False
                    break
            elif operator == "lt":
                if float(obj_value) >= float(value):
                    include = False
                    break

        if not include:
            continue

        # Build row based on schedule type
        if schedule_type == "door" and obj_type == "door":
            row = {
                "id": f"door-{obj.get('id')}",
                "objectId": obj.get("id"),
                "count": 1,
                "type": obj.get("properties", {}).get("type", "Standard"),
                "width": f"{obj.get('geometry', {}).get('width', obj.get('properties', {}).get('width', 900))}mm",
                "height": f"{obj.get('geometry', {}).get('height', obj.get('properties', {}).get('height', 2100))}mm",
                "material": obj.get("material", "-"),
                "fireRating": obj.get("properties", {}).get("fireRating", "-"),
            }
            rows.append(row)

        elif schedule_type == "window" and obj_type == "window":
            row = {
                "id": f"window-{obj.get('id')}",
                "objectId": obj.get("id"),
                "count": 1,
                "type": obj.get("properties", {}).get("type", "Fixed"),
                "width": f"{obj.get('geometry', {}).get('width', obj.get('properties', {}).get('width', 1200))}mm",
                "height": f"{obj.get('geometry', {}).get('height', obj.get('properties', {}).get('height', 1200))}mm",
                "glazing": obj.get("properties", {}).get("glazing", "Double"),
                "material": obj.get("material", "Aluminum"),
            }
            rows.append(row)

        elif schedule_type == "material":
            row = {
                "id": f"material-{obj.get('id')}",
                "objectId": obj.get("id"),
                "material": obj.get("material", "Unknown"),
                "objectCount": 1,
                "totalArea": f"{calculate_area(obj):.2f} m²",
                "totalVolume": f"{calculate_volume(obj):.3f} m³"
                if calculate_volume(obj) > 0
                else "-",
            }
            rows.append(row)

    # Calculate summary
    summary = {"totalItems": len(rows), "byType": {}}

    for row in rows:
        obj_type = row.get("type", row.get("material", "unknown"))
        summary["byType"][obj_type] = summary["byType"].get(obj_type, 0) + 1

    return jsonify(
        {
            "schedule_type": schedule_type,
            "rows": rows,
            "totalCount": len(rows),
            "generated": datetime.utcnow().isoformat(),
            "summary": summary,
        }
    )


@schedules_bp.route("/quantity-takeoff", methods=["POST"])
def quantity_takeoff():
    """
    Calculate quantity takeoff for BIM objects.

    Request body:
    {
        "objects": [list of BIM objects],
        "categories": [optional list of categories to include]
    }
    """
    data = request.get_json()

    if not data:
        return jsonify({"error": "No data provided"}), 400

    objects = data.get("objects", [])
    categories = data.get(
        "categories", ["walls", "floors", "doors", "windows", "materials"]
    )

    results = []

    # Walls
    if "walls" in categories:
        for obj in objects:
            if obj.get("type") == "wall":
                length = obj.get("geometry", {}).get(
                    "length", obj.get("properties", {}).get("length", 0)
                )
                height = obj.get("geometry", {}).get(
                    "height", obj.get("properties", {}).get("height", 0)
                )
                area = (length * height) / 1_000_000  # m²

                results.append(
                    {
                        "category": "Walls",
                        "item": obj.get("name", obj.get("id")),
                        "count": 1,
                        "unit": "m²",
                        "total": round(area, 3),
                        "unitCost": obj.get("properties", {}).get("unitCost"),
                        "totalCost": round(
                            area * obj.get("properties", {}).get("unitCost", 0), 2
                        )
                        if obj.get("properties", {}).get("unitCost")
                        else None,
                    }
                )

    # Floors/Slabs
    if "floors" in categories:
        for obj in objects:
            if obj.get("type") in ["slab", "floor", "ceiling"]:
                area = calculate_area(obj)
                volume = calculate_volume(obj)

                results.append(
                    {
                        "category": "Floors/Slabs",
                        "item": obj.get("name", obj.get("id")),
                        "count": 1,
                        "unit": "m²",
                        "total": round(area, 3),
                    }
                )

                if volume > 0:
                    results.append(
                        {
                            "category": "Concrete Volume",
                            "item": obj.get("name", obj.get("id")),
                            "count": 1,
                            "unit": "m³",
                            "total": round(volume, 3),
                        }
                    )

    # Doors
    if "doors" in categories:
        door_count = sum(1 for obj in objects if obj.get("type") == "door")
        if door_count > 0:
            results.append(
                {
                    "category": "Doors",
                    "item": "Door Count",
                    "count": door_count,
                    "unit": "ea",
                    "total": door_count,
                }
            )

    # Windows
    if "windows" in categories:
        window_count = sum(1 for obj in objects if obj.get("type") == "window")
        if window_count > 0:
            results.append(
                {
                    "category": "Windows",
                    "item": "Window Count",
                    "count": window_count,
                    "unit": "ea",
                    "total": window_count,
                }
            )

    # Materials
    if "materials" in categories:
        material_totals = {}
        for obj in objects:
            material = obj.get("material", "Unknown")
            if material not in material_totals:
                material_totals[material] = {"count": 0, "area": 0}
            material_totals[material]["count"] += 1
            material_totals[material]["area"] += calculate_area(obj)

        for material, totals in material_totals.items():
            results.append(
                {
                    "category": "Materials",
                    "item": material,
                    "count": totals["count"],
                    "unit": "ea",
                    "total": round(totals["area"], 3),
                }
            )

    return jsonify({"results": results, "generated": datetime.utcnow().isoformat()})


# ============================================================================
# Export Endpoints
# ============================================================================


@schedules_bp.route("/export/csv", methods=["POST"])
def export_csv():
    """
    Export schedule data to CSV format.

    Request body:
    {
        "schedule_type": "door" | "window" | "material",
        "headers": [list of column headers],
        "rows": [list of row data]
    }
    """
    data = request.get_json()

    if not data:
        return jsonify({"error": "No data provided"}), 400

    headers = data.get("headers", [])
    rows = data.get("rows", [])
    schedule_type = data.get("schedule_type", "schedule")

    # Build CSV
    csv_lines = [",".join(f'"{h}"' for h in headers)]
    for row in rows:
        values = [str(v) for v in row.values()]
        csv_lines.append(",".join(f'"{v}"' for v in values))

    csv_content = "\n".join(csv_lines)

    return Response(
        csv_content,
        mimetype="text/csv",
        headers={
            "Content-Disposition": f'attachment; filename="{schedule_type}_schedule.csv"'
        },
    )


@schedules_bp.route("/export/excel", methods=["POST"])
def export_excel():
    """
    Export schedule data to Excel format.

    Note: This is a placeholder. In production, you would use a library
    like openpyxl or xlsxwriter to generate proper Excel files.

    Request body:
    {
        "schedule_type": "door" | "window" | "material",
        "headers": [list of column headers],
        "rows": [list of row data],
        "title": [optional report title]
    }
    """
    data = request.get_json()

    if not data:
        return jsonify({"error": "No data provided"}), 400

    # For now, return JSON that frontend can convert
    # In production, implement proper Excel generation
    return jsonify(
        {
            "message": "Excel export placeholder - implement with openpyxl",
            "schedule_type": data.get("schedule_type"),
            "row_count": len(data.get("rows", [])),
            "note": "Install openpyxl and implement proper Excel generation",
        }
    )


@schedules_bp.route("/report", methods=["POST"])
def generate_report():
    """
    Generate full project report with schedules and quantities.

    Request body:
    {
        "project": { project metadata },
        "objects": [list of BIM objects],
        "include_schedules": [list of schedule types to include],
        "include_quantities": boolean
    """
    data = request.get_json()

    if not data:
        return jsonify({"error": "No data provided"}), 400

    project = data.get("project", {})
    objects = data.get("objects", [])
    include_schedules = data.get("include_schedules", ["door", "window", "material"])
    include_quantities = data.get("include_quantities", True)

    report = {
        "metadata": {
            "projectName": project.get("name", "Untitled Project"),
            "generated": datetime.utcnow().isoformat(),
            "author": project.get("author", "Unknown"),
        },
        "summary": {
            "totalObjects": len(objects),
            "totalMaterials": len(set(obj.get("material") for obj in objects)),
            "totalLayers": len(set(obj.get("layer") for obj in objects)),
        },
        "schedules": {},
        "quantities": None,
    }

    # Generate schedules
    for schedule_type in include_schedules:
        schedule_data = {
            "schedule_type": schedule_type,
            "rows": [],
            "totalCount": 0,
            "generated": datetime.utcnow().isoformat(),
        }

        filtered_objects = [
            obj
            for obj in objects
            if schedule_type == "material" or obj.get("type") == schedule_type
        ]

        schedule_data["totalCount"] = len(filtered_objects)
        schedule_data["rows"] = [
            {
                "id": f"{schedule_type}-{obj.get('id')}",
                "objectId": obj.get("id"),
                "type": obj.get("properties", {}).get(
                    "type", obj.get("type", "Standard")
                ),
                "dimensions": f"{obj.get('geometry', {}).get('width', '?')} × {obj.get('geometry', {}).get('height', '?')}mm",
                "material": obj.get("material", "-"),
            }
            for obj in filtered_objects
        ]

        report["schedules"][schedule_type] = schedule_data

    # Generate quantity takeoff
    if include_quantities:
        from . import quantity_takeoff

        report["quantities"] = quantity_takeoff.generate_full_quantity_takeoff(objects)

    return jsonify(report)


# ============================================================================
# Schedule Definition Endpoints
# ============================================================================


@schedules_bp.route("/definitions", methods=["GET"])
def get_schedule_definitions():
    """
    Get available schedule definitions.

    Returns list of predefined and custom schedule definitions.
    """
    definitions = [
        {
            "id": "door-schedule",
            "name": "Door Schedule",
            "type": "door",
            "description": "Complete door schedule with dimensions and materials",
            "columns": [
                {"key": "count", "header": "Count", "width": 60},
                {"key": "type", "header": "Type", "width": 100},
                {"key": "width", "header": "Width", "width": 80},
                {"key": "height", "header": "Height", "width": 80},
                {"key": "material", "header": "Material", "width": 120},
                {"key": "fireRating", "header": "Fire Rating", "width": 100},
            ],
        },
        {
            "id": "window-schedule",
            "name": "Window Schedule",
            "type": "window",
            "description": "Window schedule with sizes and glazing",
            "columns": [
                {"key": "count", "header": "Count", "width": 60},
                {"key": "type", "header": "Type", "width": 100},
                {"key": "width", "header": "Width", "width": 80},
                {"key": "height", "header": "Height", "width": 80},
                {"key": "glazing", "header": "Glazing", "width": 100},
                {"key": "material", "header": "Frame Material", "width": 120},
            ],
        },
        {
            "id": "material-schedule",
            "name": "Material Schedule",
            "type": "material",
            "description": "Material quantities by type",
            "columns": [
                {"key": "material", "header": "Material", "width": 150},
                {"key": "objectCount", "header": "Objects", "width": 80},
                {"key": "totalArea", "header": "Total Area", "width": 100},
                {"key": "totalVolume", "header": "Total Volume", "width": 100},
            ],
        },
    ]

    return jsonify({"definitions": definitions})


# ============================================================================
# Error Handlers
# ============================================================================


@schedules_bp.errorhandler(400)
def bad_request(error):
    return jsonify({"error": "Bad request"}), 400


@schedules_bp.errorhandler(500)
def internal_error(error):
    return jsonify({"error": "Internal server error"}), 500

File: frontend/src/types/schedules.ts
/**
 * Schedule Types for BIM Workbench
 * 
 * Defines schedule structures for door, window, material, and custom schedules
 */

// ============================================================================
// Schedule Definition Types
// ============================================================================

/**
 * Schedule type enumeration
 */
export type ScheduleType = 'door' | 'window' | 'material' | 'custom';

/**
 * Column definition for a schedule table
 */
export interface ScheduleColumn {
  /** Unique key for the column */
  key: string;
  /** Column header text */
  header: string;
  /** Column width in pixels */
  width: number;
  /** Function to extract value from a BIM object */
  accessor: (obj: BIMObject) => any;
  /** Optional format function for display */
  format?: (value: any) => string;
  /** Optional sort function */
  sort?: (a: any, b: any) => number;
}

/**
 * Filter definition for schedule data
 */
export interface ScheduleFilter {
  /** Property key to filter on */
  key: string;
  /** Operator for comparison */
  operator: 'equals' | 'contains' | 'gt' | 'lt' | 'gte' | 'lte';
  /** Value to compare against */
  value: any;
}

/**
 * Sort definition for schedule data
 */
export interface ScheduleSort {
  /** Column key to sort by */
  key: string;
  /** Sort direction */
  direction: 'asc' | 'desc';
}

/**
 * Schedule definition interface
 */
export interface ScheduleDefinition {
  /** Unique identifier */
  id: string;
  /** Display name */
  name: string;
  /** Schedule type */
  type: ScheduleType;
  /** Column definitions */
  columns: ScheduleColumn[];
  /** Optional filter to apply */
  filter?: (obj: BIMObject) => boolean;
  /** Optional filter definitions for UI */
  filters?: ScheduleFilter[];
  /** Default sort */
  defaultSort?: ScheduleSort;
  /** Description */
  description?: string;
}

// ============================================================================
// Schedule Data Types
// ============================================================================

/**
 * Row data for a schedule table
 */
export interface ScheduleRow {
  /** Unique row identifier */
  id: string;
  /** Reference to source BIM object */
  objectId: string;
  /** Count (usually 1, can be aggregated) */
  count: number;
  /** Object type */
  type: string;
  /** Formatted dimensions string */
  dimensions: string;
  /** Material name */
  material: string;
  /** Additional properties */
  [key: string]: any;
}

/**
 * Schedule data with metadata
 */
export interface ScheduleData {
  /** Schedule definition */
  definition: ScheduleDefinition;
  /** Row data */
  rows: ScheduleRow[];
  /** Total count */
  totalCount: number;
  /** Generated timestamp */
  generated: string;
  /** Summary statistics */
  summary?: {
    totalItems: number;
    totalArea?: number;
    totalVolume?: number;
    byType?: Record<string, number>;
  };
}

// ============================================================================
// BIM Object Types (simplified for schedules)
// ============================================================================

/**
 * Simplified BIM object for schedule calculations
 */
export interface BIMObject {
  /** Unique identifier */
  id: string;
  /** Object type */
  type: string;
  /** Display name */
  name: string;
  /** Material name */
  material: string;
  /** Layer name */
  layer: string;
  /** Object properties */
  properties: Record<string, any>;
  /** Geometry data */
  geometry?: {
    width?: number;
    height?: number;
    length?: number;
    thickness?: number;
    area?: number;
    volume?: number;
  };
  /** Position */
  position?: {
    x: number;
    y: number;
    z: number;
  };
}

// ============================================================================
// Quantity Calculation Types
// ============================================================================

/**
 * Area calculation result
 */
export interface AreaCalculation {
  objectId: string;
  objectName: string;
  objectType: string;
  area: number;
  unit: string;
}

/**
 * Volume calculation result
 */
export interface VolumeCalculation {
  objectId: string;
  objectName: string;
  objectType: string;
  volume: number;
  unit: string;
}

/**
 * Material quantity result
 */
export interface MaterialQuantity {
  materialName: string;
  objectCount: number;
  totalArea: number;
  totalVolume: number;
  unitArea: string;
  unitVolume: string;
}

/**
 * Quantity takeoff result
 */
export interface QuantityResult {
  category: string;
  item: string;
  count: number;
  unit: string;
  total: number;
  unitCost?: number;
  totalCost?: number;
}

// ============================================================================
// Predefined Schedule Definitions
// ============================================================================

/**
 * Door schedule definition
 */
export const doorSchedule: ScheduleDefinition = {
  id: 'door-schedule',
  name: 'Door Schedule',
  type: 'door',
  description: 'Complete door schedule with dimensions and materials',
  columns: [
    { key: 'count', header: 'Count', width: 60, accessor: () => 1 },
    { key: 'type', header: 'Type', width: 100, accessor: (obj) => obj.properties?.type || 'Standard' },
    { key: 'width', header: 'Width', width: 80, accessor: (obj) => `${obj.geometry?.width || obj.properties?.width || 900}mm` },
    { key: 'height', header: 'Height', width: 80, accessor: (obj) => `${obj.geometry?.height || obj.properties?.height || 2100}mm` },
    { key: 'material', header: 'Material', width: 120, accessor: (obj) => obj.material || '-' },
    { key: 'fireRating', header: 'Fire Rating', width: 100, accessor: (obj) => obj.properties?.fireRating || '-' },
  ],
  filter: (obj) => obj.type === 'door',
};

/**
 * Window schedule definition
 */
export const windowSchedule: ScheduleDefinition = {
  id: 'window-schedule',
  name: 'Window Schedule',
  type: 'window',
  description: 'Window schedule with sizes and glazing',
  columns: [
    { key: 'count', header: 'Count', width: 60, accessor: () => 1 },
    { key: 'type', header: 'Type', width: 100, accessor: (obj) => obj.properties?.type || 'Fixed' },
    { key: 'width', header: 'Width', width: 80, accessor: (obj) => `${obj.geometry?.width || obj.properties?.width || 1200}mm` },
    { key: 'height', header: 'Height', width: 80, accessor: (obj) => `${obj.geometry?.height || obj.properties?.height || 1200}mm` },
    { key: 'glazing', header: 'Glazing', width: 100, accessor: (obj) => obj.properties?.glazing || 'Double' },
    { key: 'material', header: 'Frame Material', width: 120, accessor: (obj) => obj.material || 'Aluminum' },
  ],
  filter: (obj) => obj.type === 'window',
};

/**
 * Material schedule definition
 */
export const materialSchedule: ScheduleDefinition = {
  id: 'material-schedule',
  name: 'Material Schedule',
  type: 'material',
  description: 'Material quantities by type',
  columns: [
    { key: 'material', header: 'Material', width: 150, accessor: (obj) => obj.material },
    { key: 'objectCount', header: 'Objects', width: 80, accessor: () => 1 },
    { key: 'totalArea', header: 'Total Area', width: 100, accessor: (obj) => {
      const area = obj.geometry?.area || 
        (obj.geometry?.width && obj.geometry?.height ? 
          (obj.geometry.width * obj.geometry.height) / 1000000 : 0);
      return `${area.toFixed(2)} m²`;
    }},
    { key: 'totalVolume', header: 'Total Volume', width: 100, accessor: (obj) => {
      const volume = obj.geometry?.volume || 
        (obj.geometry?.width && obj.geometry?.height && obj.geometry?.thickness ?
          (obj.geometry.width * obj.geometry.height * obj.geometry.thickness) / 1000000000 : 0);
      return volume > 0 ? `${volume.toFixed(3)} m³` : '-';
    }},
  ],
  defaultSort: { key: 'material', direction: 'asc' },
};

/**
 * All predefined schedules
 */
export const predefinedSchedules: ScheduleDefinition[] = [
  doorSchedule,
  windowSchedule,
  materialSchedule,
];

// ============================================================================
// Export/Report Types
// ============================================================================

/**
 * CSV export options
 */
export interface CSVExportOptions {
  /** Include headers */
  includeHeaders: boolean;
  /** Field delimiter */
  delimiter: string;
  /** Quote strings */
  quoteStrings: boolean;
}

/**
 * Project report structure
 */
export interface ProjectReport {
  /** Report metadata */
  metadata: {
    projectName: string;
    generated: string;
    author: string;
  };
  /** Project summary */
  summary: {
    totalObjects: number;
    totalMaterials: number;
    totalLayers: number;
    siteArea?: number;
    buildingArea?: number;
  };
  /** Schedule data */
  schedules: {
    doors: ScheduleData;
    windows: ScheduleData;
    materials: ScheduleData;
  };
  /** Quantity takeoff */
  quantities: QuantityResult[];
}

/**
 * Excel export request
 */
export interface ExcelExportRequest {
  /** Schedule definition */
  schedule: ScheduleDefinition;
  /** Schedule data */
  data: ScheduleRow[];
  /** Report title */
  title?: string;
  /** Include summary sheet */
  includeSummary?: boolean;
}
